lib/apiservers/engine/backends/commit.go

```diff
@@ -0,0 +1,472 @@
+		return nil, err
+	}
+
+	// Cleanup function for the error case
+	defer func() {
+		if err != nil {
+			os.Remove(tmpLayerFileName)
+		}
+	}()
+
+	blobSum := digest.NewDigestFromBytes(digest.SHA256, gzSum)
+	log.Debugf("container %s blob sum: %s", containerID, blobSum.String())
+	diffID := digest.NewDigestFromBytes(digest.SHA256, diffIDSum)
+	log.Debugf("container %s diff id: %s", containerID, diffID.String())
+
+	layerFile, err := os.Open(string(tmpLayerFileName))
+	if err != nil {
+		return nil, err
+	}
+	defer layerFile.Close()
+
+	decompressed, err := gzip.NewReader(layerFile)
+	if err != nil {
+		return nil, err
+	}
+	defer decompressed.Close()
+
+	// get a tar reader
+	tr := tar.NewReader(decompressed)
+
+	// iterate through tar headers to get file sizes
+	var layerSize int64
+	for {
+		tarHeader, terr := tr.Next()
+		if terr == io.EOF {
+			break
+		}
+		if terr != nil {
+			err = terr
+			return nil, err
+		}
+		layerSize += tarHeader.Size
+	}
+
+	if layerSize == 0 {
+		diffID = digest.Digest(dockerLayer.DigestSHA256EmptyTar)
+	}
+	log.Debugf("container %s size: %d", containerID, layerSize)
+
+	// Ensure the parent directory exists
+	destination := path.Join(imagec.DestinationDirectory(options), layerID)
+	err = os.MkdirAll(destination, 0755) /* #nosec */
+	if err != nil {
+		return nil, err
+	}
+
+	// Move(rename) the temporary file to its final destination
+	err = os.Rename(string(tmpLayerFileName), path.Join(destination, layerID+".tar"))
+	if err != nil {
+		return nil, err
+	}
+
+	// layer metadata
+	lm := &imagec.ImageWithMeta{
+		Image: &models.Image{
+			ID: layerID,
+		},
+		DiffID: diffID.String(),
+		Layer: imagec.FSLayer{
+			BlobSum: blobSum.String(),
+		},
+		Size: layerSize,
+	}
+	return lm, nil
+}
+
+// compressDiffToTmpFile will write stream to temp file, and return temp file name and tar file checksum, compressed file checksum
+func compressDiffToTmpFile(rc io.ReadCloser, containerID string) (string, []byte, []byte, error) {
+	defer trace.End(trace.Begin(containerID))
+	// Create a temporary file and stream the res.Body into it
+	var out *os.File
+	var gzWriter *gzip.Writer
+	var err error
+
+	cleanup := func() {
+		if gzWriter != nil {
+			gzWriter.Close()
+			gzWriter = nil
+		}
+		if out != nil {
+			out.Close()
+			if err != nil {
+				os.Remove(out.Name())
+			}
+			out = nil
+		}
+	}
+	defer cleanup()
+
+	out, err = ioutil.TempFile("", containerID)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	// compress tar file using gzip and calculate blobsum and diffID all together using multi writer
+	blobSum := sha256.New()
+	diffID := sha256.New()
+	compressedMW := io.MultiWriter(out, blobSum)
+
+	gzWriter = gzip.NewWriter(compressedMW)
+	tarMW := io.MultiWriter(gzWriter, diffID)
+	_, err = io.Copy(tarMW, rc)
+	if err != nil {
+		log.Errorf("failed to stream to file: %s", err)
+		return "", nil, nil, err
+	}
+
+	// close writer before calculate checksum
+	fileName := out.Name()
+	gzWriter.Flush()
+	cleanup()
```

Comment:
I need to flush and close writer, before call `diffID.Sum(nil), blobSum.Sum(nil)` to get correct sha256. So I set the writer to nil to avoid double close() in cleanup method